java-squid/src/main/java/org/sonar/java/se/AbstractBinaryConstraint.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import org.sonar.java.se.ConstraintManager.BooleanConstraint;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public abstract class AbstractBinaryConstraint {
+
+  public static AbstractBinaryConstraint contraintBetween(SymbolicValue v1, SymbolicValue v2, Collection<AbstractBinaryConstraint> constraints) {
+    for (AbstractBinaryConstraint constraint : constraints) {
+      AbstractBinaryConstraint c = constraint.getConstraintBetween(v1, v2, constraints);
+      if (c != null) {
+        return c;
+      }
+    }
+    return null;
+  }
+
+  protected final SymbolicValue v1;
+  protected final SymbolicValue v2;
+
+  protected AbstractBinaryConstraint(SymbolicValue v1, SymbolicValue v2) {
+    this.v1 = v1;
+    this.v2 = v2;
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (getClass().equals(obj.getClass())) {
+      return v1.equals(((AbstractBinaryConstraint) obj).v1) && v2.equals(((AbstractBinaryConstraint) obj).v2);
+    }
+    return false;
+  }
+
+  @Override
+  public int hashCode() {
+    long bits = v1.hashCode();
+    bits ^= 31 * v2.hashCode();
+    bits ^= 31 * getClass().hashCode();
+    return ((int) bits) ^ (int) (bits >> 32);
+  }
+
+  @Override
+  public String toString() {
+    StringBuilder buffer = new StringBuilder();
+    buffer.append('(');
+    buffer.append(v1);
+    buffer.append(getOperand());
+    buffer.append(v2);
+    buffer.append(')');
+    return buffer.toString();
+  }
+
+  protected AbstractBinaryConstraint getConstraintBetween(SymbolicValue v1, SymbolicValue v2, Collection<AbstractBinaryConstraint> constraints) {
+    AbstractBinaryConstraint c = null;
+    if (this.v1 == v1) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)